includes/REST_API/Embed_Controller.php

```diff
@@ -168,15 +168,47 @@ public function get_proxy_item( $request ) {
 	/**
 	 * Retrieves the story metadata for a given URL on the current site.
 	 *
-	 * @SuppressWarnings(PHPMD.NPathComplexity)
-	 *
 	 * @since 1.0.0
 	 *
 	 * @param string $url  The URL that should be inspected for metadata.
 	 *
 	 * @return array|false Story metadata if the URL does belong to the current site. False otherwise.
 	 */
 	private function get_data_from_post( $url ) {
+		$post_id = $this->url_to_post_id( $url );
+
+		if ( ! $post_id ) {
+			return false;
+		}
+
+		$post = get_post( $post_id );
+
+		if ( ! $post instanceof WP_Post ) {
+			return false;
+		}
+
+		if ( Story_Post_Type::POST_TYPE_SLUG !== $post->post_type ) {
+			return false;
+		}
+
+		return $this->get_data_from_document( $post->post_content );
+	}
+
+	/**
+	 * Examines a URL and try to determine the post ID it represents.
+	 *
+	 * Checks are supposedly from the hosted site blog.
+	 *
+	 * @SuppressWarnings(PHPMD.NPathComplexity)
+	 *
+	 * @see get_oembed_response_data_for_url
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param string $url Permalink to check.
+	 * @return int|false Post ID, or false on failure.
+	 */
+	private function url_to_post_id( $url ) {
```

Comment:
Wouldn't it be better if this method just returns the WP_Post object? The first thing that is done when is this post id is converted back to a post. Why not just cut out the middle man?